module/zfs/mmp.c

```diff
@@ -0,0 +1,465 @@
+ * uberblock has a valid ub_mmp_delay field, then the duration of the test
+ * may take longer if MMP writes were occuring less frequently than expected.
+ * Additionally, the duration is then extended by a random 25% to attempt to
+ * to detect simultaneous imports.  For example, if both partner hosts are
+ * rebooted at the same time and automatically attempt to import the pool.
+ */
+
+/*
+ * Used to control the frequency of mmp writes which are performed when the
+ * safeimport pool property is on.  This is one  factor used to determine the
+ * length of the activity check during import.
+ *
+ * The mmp write period is zfs_mmp_interval / leaf-vdevs milliseconds.  This
+ * means that on average an mmp write will be issued for each leaf vdev every
+ * zfs_mmp_interval milliseconds.  In practice, the observed period can vary
+ * with the I/O load and this observed value is the delay which is stored in
+ * the uberblock.  The minimum allowed value  is 100 ms.
+ */
+ulong_t zfs_mmp_interval = MMP_DEFAULT_INTERVAL;
+
+/*
+ * Used to control the duration of the activity test on import.  Smaller
+ * values of zfs_mmp_import_intervals will reduce the import time but increase
+ * the risk of failing to detect an active pool.  The total activity check
+ * time is never allowed to drop below one second.  A value of 0 is ignored
+ * and treated as if it was set to 1.
+ */
+uint_t zfs_mmp_import_intervals = MMP_DEFAULT_IMPORT_INTERVALS;
+
+/*
+ * Controls the behavior of the pool when mmp write failures are detected.
+ *
+ * When zfs_mmp_fail_intervals = 0 then mmp write failures are ignored.
+ * The failures will still be reported to the ZED which depending on its
+ * configuration may take action such as suspending the pool or offlining a
+ * device.
+ *
+ * When zfs_mmp_fail_intervals > 0 then sequential mmp write failures will
+ * cause the pool to be suspended.  This occurs when zfs_mmp_fail_intervals *
+ * zfs_mmp_interval milliseconds have passed since the last successful
+ * mmp write.  This guarantees the activity test will see mmp writes if the
+ * pool is imported.
+ */
+uint_t zfs_mmp_fail_intervals = MMP_DEFAULT_FAIL_INTERVALS;
+
+static void mmp_thread(spa_t *spa);
+
+void
+mmp_init(spa_t *spa)
+{
+	mmp_thread_state_t *mmp = &spa->spa_mmp;
+
+	mutex_init(&mmp->mmp_thread_lock, NULL, MUTEX_DEFAULT, NULL);
+	cv_init(&mmp->mmp_thread_cv, NULL, CV_DEFAULT, NULL);
+	mutex_init(&mmp->mmp_io_lock, NULL, MUTEX_DEFAULT, NULL);
+}
+
+void
+mmp_fini(spa_t *spa)
+{
+	mmp_thread_state_t *mmp = &spa->spa_mmp;
+
+	mutex_destroy(&mmp->mmp_thread_lock);
+	cv_destroy(&mmp->mmp_thread_cv);
+	mutex_destroy(&mmp->mmp_io_lock);
+}
+
+static void
+mmp_thread_enter(mmp_thread_state_t *mmp, callb_cpr_t *cpr)
+{
+	CALLB_CPR_INIT(cpr, &mmp->mmp_thread_lock, callb_generic_cpr, FTAG);
+	mutex_enter(&mmp->mmp_thread_lock);
+}
+
+static void
+mmp_thread_exit(mmp_thread_state_t *mmp, kthread_t **mpp, callb_cpr_t *cpr)
```

Comment:
Is providing the argument `kthread_t **mmp` necessary? This function is provided a pointer to `mmp_thread_state_t`  which contains the `kthread_t *` representing the mmp thread.